zeppelin-cluster/common/src/main/java/org/apache/zeppelin/cluster/ClusterManagerFactory.java

```diff
@@ -0,0 +1,168 @@
+  private final String defaultClusterKey;
+  private final Map<String, ClusterManager> clusterManagerMap;
+
+  private boolean initialized;
+
+  public ClusterManagerFactory(String zeppelinHome, String defaultClusterKey) {
+    this.zeppelinHome = zeppelinHome;
+    this.defaultClusterKey = defaultClusterKey;
+    this.clusterManagerMap = Maps.newHashMap();
+    this.initialized = false;
+  }
+
+  public ClusterManager getClusterManager() {
+    return getClusterManager(defaultClusterKey);
+  }
+
+  public ClusterManager getClusterManager(String key) {
+    if (!initialized) {
+      init();
+    }
+
+    if (null == key) {
+      return getClusterManager();
+    }
+
+    if (!clusterManagerMap.containsKey(key)) {
+      logger.info("Not supported. {}", key);
+      return null;
+    }
+
+    return clusterManagerMap.get(key);
+  }
+
+  synchronized void init() {
+    if (initialized) {
+      return;
+    }
+    // Find clusters inside source tree
+    logger.info("ZeppelinHome: {}", Paths.get(zeppelinHome).toAbsolutePath().toString());
+    try {
+      for (Path p : Files
+          .newDirectoryStream(Paths.get(zeppelinHome, "zeppelin-cluster"), new Filter<Path>() {
+            @Override
+            public boolean accept(Path entry) throws IOException {
+              return !entry.toString().endsWith(CLUSTER_COMMON_DIR_NAME);
+            }
+          })) {
+        String key = p.getFileName().toString();
+        List<URL> jars = Lists.newArrayList();
+        addJarsFromPath(Paths.get(p.toString(), "target"), jars);
+        addJarsFromPath(Paths.get(p.toString(), "target", "lib"), jars);
+
+        registerClusterManager(key, jars);
+      }
+
+      for (Path p : Files
+          .newDirectoryStream(Paths.get(zeppelinHome, "lib", "cluster"), new Filter<Path>() {
+            @Override
+            public boolean accept(Path entry) throws IOException {
+              return !entry.toString().endsWith(CLUSTER_COMMON_DIR_NAME);
+            }
+          })) {
+        String key = p.getFileName().toString();
+        List<URL> jars = Lists.newArrayList();
+        addJarsFromPath(p, jars);
+
+        registerClusterManager(key, jars);
+      }
+    } catch (IOException e) {
+      logger.debug("error while reading directory", e);
+    }
+
+    initialized = true;
+  }
+
+  private void registerClusterManager(String key, List<URL> jars) {
+    ClassLoader parentClassLoader = ClassLoader.getSystemClassLoader();
+    ClassLoader classLoader = new URLClassLoader(jars.toArray(new URL[0]), parentClassLoader);
+    String className = null;
+
+    try (InputStream classNameInputstream =
+             classLoader.getResourceAsStream(CLUSTER_CLASS_NAME_FILE);
+         StringWriter classNameWriter = new StringWriter()) {
+      if (null != classNameInputstream) {
+        IOUtils.copy(classNameInputstream, classNameWriter);
+        className = classNameWriter.toString();
+        Class clazz = classLoader.loadClass(className);
```

Comment:
This classloader only load ClusterManager class, but what about its dependencies ? How can they be loaded at runtime ?